### DISCUSSION TITLE
How to build the rdkit conda-forge package with tiny changes locally for myself?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [rdkit/rdkit#5590] [Q&A]
I have been using the excellent rdkit conda-forge package for a project I'm working on for a while. Recently I noticed that there is a feature missing that I require, so I forked the rdkit repo and made some [additions](https://github.com/rdkit/rdkit/compare/master...NiklasKappel:rdkit:partial_energy).

I'm not sure if my changes warrant a pull request, because the new feature is small and niche while the changes affect the force field contribution interface, so I decided to just build a modified rdkit package locally for myself. Unfortunately I have not been successful:

- Building from source using my [system compiler](https://github.com/rdkit/rdkit/blob/master/Docs/Book/Install.md#building-from-source) does not mix with my project depending on conda packages.
- Building with dependencies from [anaconda](https://github.com/rdkit/rdkit/blob/master/Docs/Book/Install.md#how-to-build-from-source-with-conda) results in an environment with an outdated Eigen version requiring me to port back parts of my projects code.
- Building with dependencies from conda-forge using the same instructions fails for me with some strange linker errors.

Then I came across the conda-forge [rdkit-feedstock](https://github.com/conda-forge/rdkit-feedstock) repo, which seems to be about building the exact package I have been using. However I don't get where to start building a package with my edits for myself, instead of pushing things through some CI/CD pipeline to the conda-forge repos.

Does anyone have some pointers for me on how I could proceed?

**Comment by bp-kelley** [comment]
Here is how I do it.  First make a copy of the recipe directory, I call it "recipe-local" or something like that.

Then change the meta.yaml source to use a path to your rdkit installation, something like:

```
source:
  path: /Users/bkelley/devel/rdkit
```

Then change the build.sh to simply echo the current directory and exit with an error after the cmake step:

```
cmake ...
echo `pwd`
exit 1
```

Now, as long as you don't run conda purge, you'll have a build directory you can run make and test in.

Don't forget that this directory is also your working .git directory now, changes to the original path will not be reflected here.

I hope this helps.

**Reply by NiklasKappel** [reply]
What am I supposed to do after editing the recipe? Run `conda build recipe-local`? That gives me a long traceback ending with:
```
ValueError: Incompatible component merge:
  - 'mpi_mpich_*'
  - 'mpi_mpich_tempest*'
```
Are there any more preparatory steps?

**Reply by bp-kelley** [reply]
It looks like this is a recent conda-build bug:   https://github.com/conda/conda/issues/11442

I'm unsure of the resolution for that.

**Reply by bp-kelley** [reply]
The chain seems to indicate that this is a python 3.10 issue,  you might try:

```
conda build --python 3.9 recipe-local
```

**Comment by bp-kelley** [comment]
I should point out that there will be a conda_build.sh in this directory that does the build and ctest, but also sets up the environment.  The downside here is that it also runs cmake every time.  (This is where the build.sh you edited above ends up, so you'll need to remove the 'exit 1' here.)

I usually make a copy of this that ONLY builds and one that ONLY tests to save time.  You can also just use the portion of that shell script that sets the environment if you like and use commands as normal.  However, ctest requires at least the RDBASE environment variable to be set correctly, so copying and editing this file is probably the way to go.

In any case, simply typing "make" in this directory will build but not test.
